### ISSUE TITLE
Problem Building GCAM 5.4

### ISSUE THREAD

**Issue by DLoughlin** [JGCRI/gcam-core#170]
I have attempted to follow the instructions to build GCAM 5.4. I also spotted the working configuration information provided by @PaulWolfram in the discussion area and have attempted to use the same versions (Boost 1.70.0, TBB 2021 2.0, Windows 10 SDK)... although I am using Visual Studio 2019 instead of 17 and Xerces C 3.2 instead of 3.1. When I attempt to compile objects-main via Visual Studio, I receive an error that appears to originate in the manage_state_variables.cpp file (line 91: "tbb::concurrent_queue<int> mThreadStateIndex;"). The error message itself is quite verbose and a bit hard to follow. It appears that the compiler is expecting another argument after "int" in "<>"(?). I've pasted the error below. Have others run into this problem and/or have some ideas about addressing it? Thank you!

Severity	Code	Description	Project	File	Line	Suppression State
Error	C2955	'std::array': use of class template requires template argument list (compiling source file ..\..\util\base\source\manage_state_variables.cpp)C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\array(430): note: see declaration of 'std::array' (compiling source file ..\..\util\base\source\manage_state_variables.cpp)E:\Projects\GLIMPSE_development\gcam-v5.4\libs\tbb\include\oneapi\tbb\detail\_concurrent_queue_base.h(447): note: while compiling class template member function 'void tbb::detail::d1::concurrent_queue_rep<T,Allocator>::assign(const tbb::detail::d1::concurrent_queue_rep<T,Allocator> &,void (__cdecl *)(int *,const void *)) with [T=int, Allocator=tbb::detail::d1::cache_aligned_allocator<int>]' (compiling source file ..\..\util\base\source\manage_state_variables.cpp)E:\Projects\GLIMPSE_development\gcam-v5.4\libs\tbb\include\oneapi\tbb\concurrent_queue.h(85): note: see reference to function template instantiation 'void tbb::detail::d1::concurrent_queue_rep<T,Allocator>::assign(const tbb::detail::d1::concurrent_queue_rep<T,Allocator> &,void (__cdecl *)(int *,const void *)) with [T=int, Allocator=tbb::detail::d1::cache_aligned_allocator<int>]' being compiled (compiling source file ..\..\util\base\source\manage_state_variables.cpp)E:\Projects\GLIMPSE_development\gcam-v5.4\libs\tbb\include\oneapi\tbb\concurrent_queue.h(112): note: see reference to class template instantiation 'tbb::detail::d1::concurrent_queue_rep<T,Allocator> with [T=int, Allocator=tbb::detail::d1::cache_aligned_allocator<int>]' being compiled (compiling source file ..\..\util\base\source\manage_state_variables.cpp)E:\Projects\GLIMPSE_development\gcam-v5.4\libs\tbb\include\oneapi\tbb\concurrent_queue.h(110): note: while compiling class template member function 'tbb::detail::d1::concurrent_queue<int,tbb::detail::d1::cache_aligned_allocator<T>>::~concurrent_queue(void) with [T=int]' (compiling source file ..\..\util\base\source\manage_state_variables.cpp)E:\Projects\GLIMPSE_development\gcam-v5.4\cvs\objects\util\base\source\manage_state_variables.cpp(94): note: see reference to function template instantiation 'tbb::detail::d1::concurrent_queue<int,tbb::detail::d1::cache_aligned_allocator<T>>::~concurrent_queue(void) with [T=int]' being compiledE:\Projects\GLIMPSE_development\gcam-v5.4\cvs\objects\util\base\source\manage_state_variables.cpp(91): note: see reference to class template instantiation 'tbb::detail::d1::concurrent_queue<int,tbb::detail::d1::cache_aligned_allocator<T>> with [T=int]' being compiled	objects-main	E:\Projects\GLIMPSE_development\gcam-v5.4\libs\tbb\include\oneapi\tbb\detail\_concurrent_queue_base.h	456

**Comment by pralitp** [comment]
Yes, we ran into this too.  Ultimately we modified the TBB include file `tbb\include\oneapi\tbb\detail_concurrent_queue_base.h` to explicitly add `this->` in front of the offending lines.

So:
```
                array[queue_idx].assign(src.array[queue_idx], *this, construct_item);
```
Becomes:
```
                this->array[queue_idx].assign(src.array[queue_idx], *this, construct_item);
```

And similarly a few lines down:
```
                array[i].clear(*this);
```
Becomes:
```
                this->array[i].clear(*this);
```

I'm not clear who's bug it is but I would guess Visual Studio as I have no idea why it would think the code is trying to create an instance of `std::array` instead of using the `array` member variables.
